File: nine-language-version/lessons/07-oop/oop.hs
-- Lesson 7: Object-Oriented Programming in Haskell
--
-- Haskell is a pure functional language and doesn't have traditional OOP.
-- However, it has concepts that achieve similar goals:
-- - Algebraic Data Types (ADTs) for data modeling
-- - Type Classes for polymorphism (different from OOP classes!)
-- - Records for bundling data
-- - Pattern matching for behavior
--
-- This demonstrates functional alternatives to OOP patterns.

{-# LANGUAGE RecordWildCards #-}

import qualified Data.Map as Map
import Data.IORef

-- ====================
-- 1. Records (like simple classes)
-- ====================

data Person = Person
    { personName :: String
    , personAge :: Int
    } deriving (Show)

-- "Constructor" is just a function
makePerson :: String -> Int -> Person
makePerson name age = Person { personName = name, personAge = age }

-- "Methods" are functions that take the record
introduce :: Person -> String
introduce p = "Hi, I'm " ++ personName p ++ ", " ++ show (personAge p) ++ " years old"

haveBirthday :: Person -> Person
haveBirthday p = p { personAge = personAge p + 1 }

-- ====================
-- 2. Type Classes (polymorphism, not OOP classes!)
-- ====================

-- Type class defines an interface
class Animal a where
    speak :: a -> String
    sleep :: a -> String
    sleep _ = "Zzz..."  -- Default implementation

-- Algebraic Data Types for different animals
data Dog = Dog
    { dogName :: String
    , dogBreed :: String
    } deriving (Show)

data Cat = Cat
    { catName :: String
    , catIndoor :: Bool
    } deriving (Show)

-- Implement the Animal type class for Dog
instance Animal Dog where
    speak d = dogName d ++ " says Woof!"
    sleep d = dogName d ++ " is sleeping... Zzz"

-- Implement the Animal type class for Cat
instance Animal Cat where
    speak c = catName c ++ " says Meow!"
    sleep c = catName c ++ " is sleeping... Zzz"

-- Additional Dog-specific behavior
fetch :: Dog -> String
fetch d = dogName d ++ " is fetching the ball!"

-- Additional Cat-specific behavior
scratch :: Cat -> String
scratch c = catName c ++ " is scratching the furniture!"

-- ====================
-- 3. Sum Types for Polymorphism
-- ====================

-- Sum type can hold any kind of animal
data AnimalType = DogAnimal Dog | CatAnimal Cat

-- Pattern matching for polymorphic behavior
animalSpeak :: AnimalType -> String
animalSpeak (DogAnimal d) = speak d
animalSpeak (CatAnimal c) = speak c

-- ====================
-- 4. Abstract Data Types with Type Classes (like abstract classes/interfaces)
-- ====================

class Shape a where
    area :: a -> Double
    perimeter :: a -> Double

    -- Default implementation using other methods
    describe :: a -> String
    describe s = shapeName s ++ ": area=" ++ show (roundTo 2 (area s))
                 ++ ", perimeter=" ++ show (roundTo 2 (perimeter s))

    shapeName :: a -> String

-- Helper function
roundTo :: Int -> Double -> Double
roundTo n x = fromIntegral (round (x * 10^n)) / 10^n

-- Concrete implementations
data Circle = Circle { circleRadius :: Double } deriving (Show)

instance Shape Circle where
    area c = pi * r * r where r = circleRadius c
    perimeter c = 2 * pi * circleRadius c
    shapeName _ = "Circle"

data Rectangle = Rectangle
    { rectWidth :: Double
    , rectHeight :: Double
    } deriving (Show)

instance Shape Rectangle where
    area r = rectWidth r * rectHeight r
    perimeter r = 2 * (rectWidth r + rectHeight r)
    shapeName _ = "Rectangle"

data Triangle = Triangle
    { triSideA :: Double
    , triSideB :: Double
    , triSideC :: Double
    } deriving (Show)

instance Shape Triangle where
    area t = sqrt (s * (s - a) * (s - b) * (s - c))
        where
            a = triSideA t
            b = triSideB t
            c = triSideC t
            s = perimeter t / 2

    perimeter t = triSideA t + triSideB t + triSideC t
    shapeName _ = "Triangle"

-- Existential type for storing different shapes together
data AnyShape = forall a. Shape a => AnyShape a

instance Shape AnyShape where
    area (AnyShape s) = area s
    perimeter (AnyShape s) = perimeter s
    shapeName (AnyShape s) = shapeName s

-- ====================
-- 5. Encapsulation with Abstract Data Types
-- ====================

-- Only export the type and specific functions, not the constructor
-- This provides encapsulation (in a real module)

data BankAccount = BankAccount
    { accountNumber :: String
    , balance :: Double
    , transactions :: [String]
    } deriving (Show)

newAccount :: String -> Double -> BankAccount
newAccount accNum initialBalance =
    BankAccount accNum initialBalance []

deposit :: Double -> BankAccount -> Maybe BankAccount
deposit amount acc
    | amount > 0 = Just acc
        { balance = balance acc + amount
        , transactions = transactions acc ++ ["Deposit: +$" ++ show amount]
        }
    | otherwise = Nothing

withdraw :: Double -> BankAccount -> Maybe BankAccount
withdraw amount acc
    | amount > 0 && amount <= balance acc = Just acc
        { balance = balance acc - amount
        , transactions = transactions acc ++ ["Withdrawal: -$" ++ show amount]
        }
    | otherwise = Nothing

getBalance :: BankAccount -> Double
getBalance = balance

getTransactions :: BankAccount -> [String]
getTransactions = transactions

-- ====================
-- 6. "Smart Constructors" and Phantom Types
-- ====================

data Temperature = Temperature { celsius :: Double } deriving (Show)

fromCelsius :: Double -> Temperature
fromCelsius c = Temperature c

fromFahrenheit :: Double -> Temperature
fromFahrenheit f = Temperature ((f - 32) * 5 / 9)

fromKelvin :: Double -> Temperature
fromKelvin k = Temperature (k - 273.15)

isFreezing :: Temperature -> Bool
isFreezing t = celsius t <= 0

toFahrenheit :: Temperature -> Double
toFahrenheit t = celsius t * 9 / 5 + 32

toKelvin :: Temperature -> Double
toKelvin t = celsius t + 273.15

-- ====================
-- 7. Composition (preferred in FP!)
-- ====================

data Engine = Engine
    { horsepower :: Int
    , running :: Bool
    } deriving (Show)

newEngine :: Int -> Engine
newEngine hp = Engine hp False

startEngine :: Engine -> Engine
startEngine e = e { running = True }

stopEngine :: Engine -> Engine
stopEngine e = e { running = False }

data Car = Car
    { brand :: String
    , model :: String
    , engine :: Engine  -- Composition!
    } deriving (Show)

newCar :: String -> String -> Int -> Car
newCar b m hp = Car b m (newEngine hp)

startCar :: Car -> (String, Car)
startCar c =
    let newEngine = startEngine (engine c)
        msg = brand c ++ " " ++ model c ++ ": Engine starting... "
              ++ show (horsepower newEngine) ++ "hp engine now running"
    in (msg, c { engine = newEngine })

stopCar :: Car -> (String, Car)
stopCar c =
    let newEngine = stopEngine (engine c)
        msg = brand c ++ " " ++ model c ++ ": Engine stopped"
    in (msg, c { engine = newEngine })

-- ====================
-- 8. Type Classes for Operator-like Behavior
-- ====================

data Point = Point
    { pointX :: Double
    , pointY :: Double
    } deriving (Show, Eq)

-- Point is an instance of Num, so we can use +, -, etc.
instance Num Point where
    Point x1 y1 + Point x2 y2 = Point (x1 + x2) (y1 + y2)
    Point x1 y1 - Point x2 y2 = Point (x1 - x2) (y1 - y2)
    Point x1 y1 * Point x2 y2 = Point (x1 * x2) (y1 * y2)  -- Element-wise
    abs (Point x y) = Point (abs x) (abs y)
    signum (Point x y) = Point (signum x) (signum y)
    fromInteger n = Point (fromInteger n) (fromInteger n)

distanceFromOrigin :: Point -> Double
distanceFromOrigin (Point x y) = sqrt (x * x + y * y)

-- ====================
-- 9. Multiple Type Class Constraints (like multiple interfaces)
-- ====================

class Flyable a where
    fly :: a -> String

class Swimmable a where
    swim :: a -> String

data Duck = Duck { duckName :: String } deriving (Show)

instance Animal Duck where
    speak d = duckName d ++ " says Quack!"

instance Flyable Duck where
    fly d = duckName d ++ " is flying through the air"

instance Swimmable Duck where
    swim d = duckName d ++ " is swimming in water"

-- Function requiring multiple constraints
showDuckAbilities :: (Animal a, Flyable a, Swimmable a) => a -> IO ()
showDuckAbilities d = do
    putStrLn $ "   " ++ speak d
    putStrLn $ "   " ++ fly d
    putStrLn $ "   " ++ swim d

-- ====================
-- 10. Design Patterns in Functional Style
-- ====================

-- Factory Pattern (just a function!)
createAnimal :: String -> String -> AnimalType
createAnimal "dog" name = DogAnimal (Dog name "Mixed")
createAnimal "cat" name = CatAnimal (Cat name True)
createAnimal _ name = DogAnimal (Dog name "Unknown")

-- Builder Pattern (using function composition)
data PersonBuilder = PersonBuilder
    { builderName :: Maybe String
    , builderAge :: Maybe Int
    }

emptyBuilder :: PersonBuilder
emptyBuilder = PersonBuilder Nothing Nothing

setName :: String -> PersonBuilder -> PersonBuilder
setName n pb = pb { builderName = Just n }

setAge :: Int -> PersonBuilder -> PersonBuilder
setAge a pb = pb { builderAge = Just a }

buildPerson :: PersonBuilder -> Maybe Person
buildPerson pb = do
    n <- builderName pb
    a <- builderAge pb
    return $ Person n a

-- ====================
-- Main Function
-- ====================

main :: IO ()
main = do
    putStrLn "=== Object-Oriented Programming in Haskell ===\n"

    -- 1. Basic record
    putStrLn "1. Basic Record (like a simple class):"
    let alice = makePerson "Alice" 30
    putStrLn $ "   " ++ introduce alice
    let alice' = haveBirthday alice
    putStrLn $ "   After birthday: age = " ++ show (personAge alice')

    -- 2. Type classes and polymorphism
    putStrLn "\n2. Type Classes and Polymorphism:"
    let buddy = Dog "Buddy" "Golden Retriever"
    let whiskers = Cat "Whiskers" True
    let maxDog = Dog "Max" "German Shepherd"

    putStrLn $ "   " ++ speak buddy
    putStrLn $ "   " ++ speak whiskers
    putStrLn $ "   " ++ speak maxDog

    putStrLn $ "   " ++ fetch buddy
    putStrLn $ "   " ++ scratch whiskers

    -- 3. Polymorphism with sum types
    putStrLn "\n   Using sum types for polymorphism:"
    let animals = [DogAnimal buddy, CatAnimal whiskers, DogAnimal maxDog]
    mapM_ (\a -> putStrLn $ "   " ++ animalSpeak a) animals

    -- 4. Shapes
    putStrLn "\n3. Shapes (Type Classes):"
    let circle = Circle 5.0
    let rectangle = Rectangle 4.0 6.0
    let triangle = Triangle 3.0 4.0 5.0

    putStrLn $ "   " ++ describe circle
    putStrLn $ "   " ++ describe rectangle
    putStrLn $ "   " ++ describe triangle

    -- 5. Encapsulation
    putStrLn "\n4. Encapsulation (Bank Account):"
    let account = newAccount "ACC001" 1000.0
    putStrLn $ "   Initial balance: $" ++ show (getBalance account)
    case deposit 500.0 account of
        Just acc1 -> do
            putStrLn $ "   After deposit: $" ++ show (getBalance acc1)
            case withdraw 200.0 acc1 of
                Just acc2 -> do
                    putStrLn $ "   After withdrawal: $" ++ show (getBalance acc2)
                    putStrLn $ "   Transactions: " ++ show (getTransactions acc2)
                Nothing -> putStrLn "   Withdrawal failed"
        Nothing -> putStrLn "   Deposit failed"

    -- 6. Smart constructors
    putStrLn "\n5. Smart Constructors (Temperature):"
    let temp1 = fromCelsius 0
    let temp2 = fromFahrenheit 32
    let temp3 = fromKelvin 273.15

    putStrLn $ "   0°C = " ++ show (roundTo 1 (toFahrenheit temp1)) ++ "°F"
    putStrLn $ "   32°F = " ++ show (roundTo 1 (celsius temp2)) ++ "°C"
    putStrLn $ "   273.15K = " ++ show (roundTo 1 (celsius temp3)) ++ "°C"
    putStrLn $ "   Is 0°C freezing? " ++ show (isFreezing temp1)

    -- 7. Composition
    putStrLn "\n6. Composition:"
    let car = newCar "Toyota" "Camry" 200
    let (startMsg, car') = startCar car
    putStrLn $ "   " ++ startMsg
    let (stopMsg, car'') = stopCar car'
    putStrLn $ "   " ++ stopMsg

    -- 8. Operator overloading with type classes
    putStrLn "\n7. Operator Overloading (Type Classes):"
    let p1 = Point 3 4
    let p2 = Point 1 2
    putStrLn $ "   p1 = " ++ show p1
    putStrLn $ "   p2 = " ++ show p2
    putStrLn $ "   p1 + p2 = " ++ show (p1 + p2)
    putStrLn $ "   p1 - p2 = " ++ show (p1 - p2)
    putStrLn $ "   p1 == p2? " ++ show (p1 == p2)
    putStrLn $ "   p1 distance from origin: " ++ show (roundTo 2 (distanceFromOrigin p1))

    -- 9. Multiple type class constraints
    putStrLn "\n8. Multiple Type Class Constraints:"
    let duck = Duck "Donald"
    showDuckAbilities duck

    -- 10. Factory pattern
    putStrLn "\n9. Factory Pattern:"
    let dog = createAnimal "dog" "Rover"
    let cat = createAnimal "cat" "Mittens"
    putStrLn $ "   " ++ animalSpeak dog
    putStrLn $ "   " ++ animalSpeak cat

    -- 11. Builder pattern
    putStrLn "\n10. Builder Pattern:"
    case buildPerson $ setAge 25 $ setName "Bob" emptyBuilder of
        Just bob -> putStrLn $ "   " ++ introduce bob
        Nothing -> putStrLn "   Failed to build person"

    putStrLn "\n11. Key Takeaway:"
    putStrLn "   Haskell doesn't have OOP, but achieves similar goals through:"
    putStrLn "   - Type classes for polymorphism"
    putStrLn "   - ADTs for modeling data"
    putStrLn "   - Immutability and pure functions"
    putStrLn "   - Composition over inheritance"

File: nine-language-version/lessons/07-oop/oop.py
#!/usr/bin/env python3
"""
Lesson 7: Object-Oriented Programming in Python

Python is a multi-paradigm language with full OOP support.
Everything in Python is an object!
"""

from abc import ABC, abstractmethod
import math

# ====================
# 1. Basic Class
# ====================

class Person:
    """Basic class demonstrating encapsulation"""

    def __init__(self, name, age):
        self.name = name
        self._age = age  # "Protected" (convention only)

    def introduce(self):
        return f"Hi, I'm {self.name}, {self._age} years old"

    def have_birthday(self):
        self._age += 1

    def get_age(self):
        return self._age


# ====================
# 2. Inheritance
# ====================

class Animal:
    """Base class for animals"""

    def __init__(self, name, species):
        self.name = name
        self.species = species

    def speak(self):
        return "Some generic animal sound"

    def sleep(self):
        return f"{self.name} is sleeping... Zzz"


class Dog(Animal):
    """Dog inherits from Animal"""

    def __init__(self, name, breed):
        super().__init__(name, "Canine")
        self.breed = breed

    def speak(self):  # Override
        return f"{self.name} says Woof!"

    def fetch(self):
        return f"{self.name} is fetching the ball!"


class Cat(Animal):
    """Cat inherits from Animal"""

    def __init__(self, name, indoor=True):
        super().__init__(name, "Feline")
        self.indoor = indoor

    def speak(self):  # Override
        return f"{self.name} says Meow!"

    def scratch(self):
        return f"{self.name} is scratching the furniture!"


# ====================
# 3. Abstract Classes
# ====================

class Shape(ABC):
    """Abstract base class for shapes"""

    @abstractmethod
    def area(self):
        """Calculate area - must be implemented by subclasses"""
        pass

    @abstractmethod
    def perimeter(self):
        """Calculate perimeter - must be implemented by subclasses"""
        pass

    def describe(self):
        """Concrete method available to all shapes"""
        return f"{self.__class__.__name__}: area={self.area():.2f}, perimeter={self.perimeter():.2f}"


class Circle(Shape):
    """Circle implementation"""

    def __init__(self, radius):
        self.radius = radius

    def area(self):
        return math.pi * self.radius ** 2

    def perimeter(self):
        return 2 * math.pi * self.radius


class Rectangle(Shape):
    """Rectangle implementation"""

    def __init__(self, width, height):
        self.width = width
        self.height = height

    def area(self):
        return self.width * self.height

    def perimeter(self):
        return 2 * (self.width + self.height)


class Triangle(Shape):
    """Triangle implementation"""

    def __init__(self, side_a, side_b, side_c):
        self.side_a = side_a
        self.side_b = side_b
        self.side_c = side_c

    def area(self):
        # Heron's formula
        s = self.perimeter() / 2
        return math.sqrt(s * (s - self.side_a) * (s - self.side_b) * (s - self.side_c))

    def perimeter(self):
        return self.side_a + self.side_b + self.side_c


# ====================
# 4. Encapsulation (Bank Account)
# ====================

class BankAccount:
    """Demonstrates encapsulation and data hiding"""

    def __init__(self, account_number, initial_balance=0):
        self.account_number = account_number
        self.__balance = initial_balance  # Private attribute
        self.__transactions = []

    def deposit(self, amount):
        if amount > 0:
            self.__balance += amount
            self.__transactions.append(f"Deposit: +${amount}")
            return True
        return False

    def withdraw(self, amount):
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            self.__transactions.append(f"Withdrawal: -${amount}")
            return True
        return False

    def get_balance(self):
        return self.__balance

    def get_transaction_history(self):
        return self.__transactions.copy()


# ====================
# 5. Class and Static Methods
# ====================

class Temperature:
    """Demonstrates class and static methods"""

    # Class variable
    conversion_count = 0

    def __init__(self, celsius):
        self.celsius = celsius

    @classmethod
    def from_fahrenheit(cls, fahrenheit):
        """Factory method using class method"""
        cls.conversion_count += 1
        celsius = (fahrenheit - 32) * 5 / 9
        return cls(celsius)

    @classmethod
    def from_kelvin(cls, kelvin):
        """Factory method using class method"""
        cls.conversion_count += 1
        celsius = kelvin - 273.15
        return cls(celsius)

    @staticmethod
    def is_freezing(celsius):
        """Utility method that doesn't need instance or class"""
        return celsius <= 0

    def to_fahrenheit(self):
        return (self.celsius * 9 / 5) + 32

    def to_kelvin(self):
        return self.celsius + 273.15


# ====================
# 6. Composition
# ====================

class Engine:
    """Component class"""

    def __init__(self, horsepower):
        self.horsepower = horsepower
        self.running = False

    def start(self):
        self.running = True
        return f"Engine starting... {self.horsepower}hp engine now running"

    def stop(self):
        self.running = False
        return "Engine stopped"


class Car:
    """Uses composition - Car HAS-AN Engine"""

    def __init__(self, brand, model, horsepower):
        self.brand = brand
        self.model = model
        self.engine = Engine(horsepower)  # Composition!

    def start(self):
        return f"{self.brand} {self.model}: {self.engine.start()}"

    def stop(self):
        return f"{self.brand} {self.model}: {self.engine.stop()}"


# ====================
# 7. Special Methods (Magic Methods)
# ====================

class Point:
    """Demonstrates operator overloading with magic methods"""

    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        """String representation"""
        return f"Point({self.x}, {self.y})"

    def __repr__(self):
        """Official representation"""
        return f"Point({self.x}, {self.y})"

    def __eq__(self, other):
        """Equality operator =="""
        if not isinstance(other, Point):
            return False
        return self.x == other.x and self.y == other.y

    def __add__(self, other):
        """Addition operator +"""
        return Point(self.x + other.x, self.y + other.y)

    def __sub__(self, other):
        """Subtraction operator -"""
        return Point(self.x - other.x, self.y - other.y)

    def distance_from_origin(self):
        return math.sqrt(self.x ** 2 + self.y ** 2)


# ====================
# 8. Property Decorators
# ====================

class Circle2:
    """Demonstrates @property decorator"""

    def __init__(self, radius):
        self._radius = radius

    @property
    def radius(self):
        """Getter for radius"""
        return self._radius

    @radius.setter
    def radius(self, value):
        """Setter with validation"""
        if value < 0:
            raise ValueError("Radius cannot be negative")
        self._radius = value

    @property
    def diameter(self):
        """Computed property"""
        return self._radius * 2

    @property
    def area(self):
        """Computed property"""
        return math.pi * self._radius ** 2


# ====================
# 9. Design Pattern: Singleton
# ====================

class Singleton:
    """Singleton pattern - only one instance allowed"""

    _instance = None

    def __new__(cls):
        if cls._instance is None:
            cls._instance = super().__new__(cls)
            cls._instance.initialized = False
        return cls._instance

    def __init__(self):
        if not self.initialized:
            self.data = []
            self.initialized = True


# ====================
# 10. Design Pattern: Factory
# ====================

class AnimalFactory:
    """Factory pattern for creating animals"""

    @staticmethod
    def create_animal(animal_type, name):
        if animal_type.lower() == "dog":
            return Dog(name, "Mixed")
        elif animal_type.lower() == "cat":
            return Cat(name)
        else:
            return Animal(name, "Unknown")


# ====================
# Tests and Examples
# ====================

def main():
    print("=== Object-Oriented Programming in Python ===\n")

    # 1. Basic class
    print("1. Basic Class:")
    alice = Person("Alice", 30)
    print(f"   {alice.introduce()}")
    alice.have_birthday()
    print(f"   After birthday: age = {alice.get_age()}")

    # 2. Inheritance and Polymorphism
    print("\n2. Inheritance and Polymorphism:")
    animals = [
        Dog("Buddy", "Golden Retriever"),
        Cat("Whiskers", indoor=True),
        Dog("Max", "German Shepherd")
    ]

    for animal in animals:
        print(f"   {animal.speak()}")

    print(f"   {animals[0].fetch()}")
    print(f"   {animals[1].scratch()}")

    # 3. Abstract classes and shapes
    print("\n3. Abstract Classes (Shapes):")
    shapes = [
        Circle(5),
        Rectangle(4, 6),
        Triangle(3, 4, 5)
    ]

    for shape in shapes:
        print(f"   {shape.describe()}")

    # 4. Encapsulation (Bank Account)
    print("\n4. Encapsulation (Bank Account):")
    account = BankAccount("ACC001", 1000)
    print(f"   Initial balance: ${account.get_balance()}")
    account.deposit(500)
    print(f"   After deposit: ${account.get_balance()}")
    account.withdraw(200)
    print(f"   After withdrawal: ${account.get_balance()}")
    print(f"   Transactions: {account.get_transaction_history()}")

    # 5. Class and static methods
    print("\n5. Class and Static Methods:")
    temp1 = Temperature(0)
    temp2 = Temperature.from_fahrenheit(32)
    temp3 = Temperature.from_kelvin(273.15)

    print(f"   0°C = {temp1.to_fahrenheit():.1f}°F")
    print(f"   32°F = {temp2.celsius:.1f}°C")
    print(f"   273.15K = {temp3.celsius:.1f}°C")
    print(f"   Is 0°C freezing? {Temperature.is_freezing(0)}")
    print(f"   Conversions made: {Temperature.conversion_count}")

    # 6. Composition
    print("\n6. Composition:")
    car = Car("Toyota", "Camry", 200)
    print(f"   {car.start()}")
    print(f"   {car.stop()}")

    # 7. Special methods
    print("\n7. Special Methods (Operator Overloading):")
    p1 = Point(3, 4)
    p2 = Point(1, 2)
    print(f"   p1 = {p1}")
    print(f"   p2 = {p2}")
    print(f"   p1 + p2 = {p1 + p2}")
    print(f"   p1 - p2 = {p1 - p2}")
    print(f"   p1 == p2? {p1 == p2}")
    print(f"   p1 distance from origin: {p1.distance_from_origin():.2f}")

    # 8. Properties
    print("\n8. Properties:")
    circle = Circle2(5)
    print(f"   Radius: {circle.radius}")
    print(f"   Diameter: {circle.diameter}")
    print(f"   Area: {circle.area:.2f}")
    circle.radius = 10
    print(f"   After setting radius to 10:")
    print(f"   Diameter: {circle.diameter}")
    print(f"   Area: {circle.area:.2f}")

    # 9. Singleton
    print("\n9. Singleton Pattern:")
    s1 = Singleton()
    s2 = Singleton()
    print(f"   s1 is s2? {s1 is s2}")
    s1.data.append("item1")
    print(f"   s1.data: {s1.data}")
    print(f"   s2.data: {s2.data}")

    # 10. Factory
    print("\n10. Factory Pattern:")
    dog = AnimalFactory.create_animal("dog", "Rover")
    cat = AnimalFactory.create_animal("cat", "Mittens")
    print(f"   {dog.speak()}")
    print(f"   {cat.speak()}")


if __name__ == "__main__":
    main()
